CtrlVAF/CtrlVAF.Core/EventHandlers/EventDispatcher.cs

```diff
@@ -0,0 +1,136 @@
+
+    public class EventDispatcher : Dispatcher
+    {
+
+        /// <summary>
+        /// Typical usage of this class would be inside an event handler method inside the vault application base class.
+        /// Once called, the dispatcher will locate any ICommandHandlers using the same TCommand interface and invoke their handle method.
+        /// </summary>
+        /// <param name="command">A command of type TCommand that has inherited from <see cref="Commands.IEventCommand{T}"/></param>
+        /// <param name="throwExceptions">Whether or not to stop executing ICommandHandlers upon exceptions and throw the exception</param>
+        /// <param name="exceptionHandler">An exception handler to pass along or handle any ICommandHandler exceptions</param>
+        public EventDispatcher()
+        {
+        }
+
+        /// <inheritdoc/>
+        public override void Dispatch(params ICtrlVAFCommand[] commands)
+        {
+            IncludeAssemblies(Assembly.GetCallingAssembly());
+
+            var concreteTypes = GetTypes(commands);
+
+            HandleConcreteTypes(concreteTypes, commands);
+        }
+
+        protected internal override IEnumerable<Type> GetTypes(params ICtrlVAFCommand[] commands)
+        {
+            // Instantiate a handlerType according to the TCommand type provided
+            Type handlerType = typeof(IEventHandler<>);
+            List<Type> dispatchableHandlerTypes = new List<Type>();
+
+            foreach (ICtrlVAFCommand command in commands)
+            {
+                Type commandType = command.GetType();
+
+                //If the concrete types have already been retrieved and cached before, simply handle those
+                if (TypeCache.TryGetValue(commandType, out var cachedTypes))
+                {
+                    dispatchableHandlerTypes.AddRange(cachedTypes.Distinct());
+                    continue;
+                }
+
+                List<Type> handlerTypes = new List<Type>();
+
+                foreach (Assembly assembly in Assemblies)
+                {
+                    var types = assembly.GetTypes();
+
+                    foreach (Type type in types)
+                    {
+                        if (type.IsClass)
+                        {
+                            var commandHandlerInterfaces = type.GetInterfaces().Where(t =>
+                            t.IsGenericType &&
+                            t.GetGenericTypeDefinition() == handlerType &&
+                            t.GenericTypeArguments[0] == commandType);
+
+                            if (commandHandlerInterfaces.Any())
+                                handlerTypes.Add(type);
+                        }
+                    }
+                }
+
+                TypeCache.TryAdd(commandType, handlerTypes);
+
+                dispatchableHandlerTypes.AddRange(handlerTypes);
+            }
+
+            return dispatchableHandlerTypes;
```

Comment:
I suppose that here there is a slim chance of duplicate handlerTypes being returned?
As multiple commands of the same type could be dispatched in the beginning.
Perhaps, a distinct call should be placed here?

Reply:
I prefer keeping a log of handled command types then. It seems more efficient than adding duplicates and then removing them.